src/computability/language.lean

```diff
@@ -151,4 +151,98 @@ begin
     finish }
 end
 
+lemma le_iff (l m : language α) : l ≤ m ↔ l + m = m :=
+begin
+  rw [set.le_eq_subset, add_def],
+  split,
+  { intro h,
+    ext x,
+    simp only [set.mem_union_eq, or_iff_right_iff_imp],
+    tauto },
+  { intro h,
+    rw ←h,
+    exact set.subset_union_left l m }
+end
+
+lemma one_add_lang_lang_star_le_lang_star (l : language α) : 1 + l*l.star ≤ l.star :=
+begin
+  rw [set.le_eq_subset, one_def, add_def, mul_def, star_def],
+  intros x hx,
+  simp only [exists_and_distrib_left, set.mem_image2, set.image_prod, set.mem_union_eq,
+    set.mem_set_of_eq] at hx,
+  cases hx,
+  { rw [set.mem_singleton_iff] at hx,
+    subst hx,
+    use [[]],
+    split,
+    refl,
+    tauto },
+  { rcases hx with ⟨ a, ha, b, ⟨ ⟨ S, hb, hS ⟩, hx ⟩ ⟩,
+    use a :: S,
+    split,
+    finish,
+    rintros y (hy|hy);
+    finish }
+end
+
+lemma one_add_lang_star_lang_le_lang_star (l : language α) : 1 + l.star*l ≤ l.star :=
+begin
+  rw [set.le_eq_subset, one_def, add_def, mul_def, star_def],
+  intros x hx,
+  simp only [exists_and_distrib_left, set.mem_image2, set.image_prod, set.mem_union_eq,
+    set.mem_set_of_eq] at hx,
+  cases hx,
+  { rw [set.mem_singleton_iff] at hx,
+    subst hx,
+    use [[]],
+    split,
+    refl,
+    tauto },
+  { rcases hx with ⟨ b, ⟨ S, hb, hS ⟩, a, ha, hx ⟩,
+    use S ++ [a],
+    split,
+    finish,
+    simp only [list.mem_append, list.mem_singleton],
+    rintros y (hy|hy);
+    finish }
+end
+
+lemma star_mul_le_right_of_mul_le_right (l m : language α) : l*m ≤ m → l.star*m ≤ m :=
```

Comment:
I'm a bit worried that the proofs of these very basic facts are already so complicated.

Maybe the following would work better:
1. Derive `complete_lattice (language α)`.
2. Prove `le_mul_congr : l1 ≤ m1 → l2 ≤ m2 → l1*l2 ≤ m1*l2`, etc.
2. Prove `supr_mul : (⨆ i, l i) * m = ⨆ i, l i * m` and permutations, and also for `+`
3. Prove `star_eq_supr_pow : l.star = ⨆ i : ℕ, l ^ n`.

Then this proof becomes:
```lean
begin
 intro h,
 rw [star_eq_supr_pow, supr_mul],
 apply supr_le, intro n,
 induction n,
 { simp },
 -- ih : l^n ≤ m
 -- ⊢ l^(n+1) * m ≤ m
 rw mul_assoc (l^n) l m,
 exact le_trans (le_mul_congr (le_refl _) h) ih,
end
```